### DISCUSSION TITLE
Key based memoization with automatic freeing based on subsciber count

### DISCUSSION THREAD

**Discussion by ssadler** [pmndrs/jotai#2012] [Ideas]
In my app, I need to load a sort of tree of resources for my react components, and in some cases each one might be loaded many times, as well as loaded and unloaded quite quickly.

I wanted to deduplicate the requests, as well as automatically free when there are no more subscribers, so I came up with something like:

```js
const fooLoader = loader({
  key: "someFoo",
  async get() { return await fetch(...) }
})

const barLoader = loader({
  key: "someBar",
  async get(use) {
    return await use(fooLoader) * 2
  }
})
```

The trick is, that it has a global cache, so anytime a loader subscribed to, it checks if it's already in the cache with the same key. And when there are no more subscibers, it deletes itself from the cache. To illustrate:

```js
subscribe(listener) {
  if (loaderCache[this.key]) return loaderCache[this.key]
  ...
}
unsubscribe() {
  if (this.subscribers.length == 0) {
    this.stop()
    delete loadedCache[this.key]
  }
}
```

The other neat feature is a configurable timeout before freeing from the cache.

Would this be easy to implement using Jotai?

**Comment by dai-shi** [comment]
Jotai is basically a WeakMap, so you just need to control the lifetime of atoms by yourself.

I think you can do what you want with your own version of atomFamily.
Here's the basic atomFamily impl. https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035
```js
// simplified code
const atomFamily = (initializeAtom) => {
  const cache = new Map();
  return (param) => {
    if (!cache.has(param)) {
      cache.set(param, initializeAtom(param));
    }
    return cache.get(param);
  };
};
```

I think you can add the (un)subscribe capability there.
